File: backend/app/api/v1/courses.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import List, Optional
from ...core.database import get_db
from ...core.security import get_current_user, get_current_instructor_or_admin
from ...models.user import User
from ...models.course import Course, DifficultyLevel
from ...schemas.course import CourseCreate, CourseUpdate, CourseResponse

router = APIRouter(tags=["Courses"])

@router.post("/courses", response_model=CourseResponse, status_code=status.HTTP_201_CREATED)
def create_course(
    course_data: CourseCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_instructor_or_admin)
):
    new_course = Course(
        **course_data.model_dump(),
        instructor_id=current_user.id
    )
    db.add(new_course)
    db.commit()
    db.refresh(new_course)
    return new_course

@router.get("/courses", response_model=List[CourseResponse])
def list_courses(
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    category: Optional[str] = None,
    difficulty: Optional[DifficultyLevel] = None,
    search: Optional[str] = None,
    db: Session = Depends(get_db)
):
    # query = db.query(Course).filter(Course.is_published == True)   #AJUN COURCE ACESS NAHI 
    query = db.query(Course)   #COURCE ACESS
    if category:
        query = query.filter(Course.category == category)
    if difficulty:
        query = query.filter(Course.difficulty == difficulty)
    if search:
        query = query.filter(
            Course.title.ilike(f"%{search}%") | 
            Course.description.ilike(f"%{search}%")
        )
    
    courses = query.offset(skip).limit(limit).all()
    return courses

@router.get("/courses/{course_id}", response_model=CourseResponse)
def get_course(course_id: int, db: Session = Depends(get_db)):
    course = db.query(Course).filter(Course.id == course_id).first()
    if not course:
        raise HTTPException(status_code=404, detail="Course not found")
    return course

@router.put("/courses/{course_id}", response_model=CourseResponse)
def update_course(
    course_id: int,
    course_update: CourseUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_instructor_or_admin)
):
    course = db.query(Course).filter(Course.id == course_id).first()
    if not course:
        raise HTTPException(status_code=404, detail="Course not found")
    
    if current_user.role != "admin" and course.instructor_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized")
    
    for field, value in course_update.model_dump(exclude_unset=True).items():
        setattr(course, field, value)
    
    db.commit()
    db.refresh(course)
    return course

@router.delete("/courses/{course_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_course(
    course_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_instructor_or_admin)
):
    course = db.query(Course).filter(Course.id == course_id).first()
    if not course:
        raise HTTPException(status_code=404, detail="Course not found")
    
    if current_user.role != "admin" and course.instructor_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized")
    
    db.delete(course)
    db.commit()
    return {"message": "Course deleted successfully"}

File: frontend/src/api/courses.ts
import api from './index';

// ✅ Complete Course interface with progress
export interface Course {
  id: number;
  title: string;
  description: string;
  category: string;
  difficulty: string;
  duration: number;
  price: number;
  instructor_id: number;
  is_published: boolean;
  cover_image?: string;
  created_at: string;
  updated_at?: string;
  progress?: number;  // ✅ Optional progress property
}

// ✅ Module Interface
export interface Module {
  id: number;
  course_id: number;
  title: string;
  description: string;
  order: number;
  lessons?: Lesson[];
  created_at: string;
  updated_at: string;
}

// ✅ Lesson Interface
export interface Lesson {
  id: number;
  module_id: number;
  title: string;
  description: string;
  concept: string;
  example: string;
  interview_questions: string;
  duration: number;
  order: number;
  is_completed?: boolean;
  created_at: string;
  updated_at: string;
}

export interface CourseCreate {
  title: string;
  description: string;
  category: string;
  difficulty: 'beginner' | 'intermediate' | 'advanced';
  duration: number;
  price: number;
}

export const getCourses = async (params?: {
  skip?: number;
  limit?: number;
  category?: string;
  difficulty?: string;
  search?: string;
}): Promise<Course[]> => {
  const response = await api.get<Course[]>('/courses', { params });
  return response.data;
};

export const getCourse = async (id: number): Promise<Course> => {
  const response = await api.get<Course>(`/courses/${id}`);
  return response.data;
};

export const getCourseModules = async (courseId: number): Promise<Module[]> => {
  const response = await api.get<Module[]>(`/courses/${courseId}/modules`);
  return response.data;
};

export const createCourse = async (data: CourseCreate): Promise<Course> => {
  const response = await api.post<Course>('/courses', data);
  return response.data;
};

export const updateCourse = async (id: number, data: Partial<CourseCreate>): Promise<Course> => {
  const response = await api.put<Course>(`/courses/${id}`, data);
  return response.data;
};

export const deleteCourse = async (id: number): Promise<void> => {
  await api.delete(`/courses/${id}`);
};